glib-macros/src/clone.rs

```diff
@@ -1,952 +1,487 @@
-            _ => false,
+        if !return_closure.inputs.is_empty() {
+            return Err(syn::Error::new_spanned(
+                &return_closure,
+                "Default return value closure must not have parameters",
+            ));
         }
-    }
-}
 
-fn is_punct(elem: &TokenTree, punct: &str) -> bool {
-    match elem {
-        TokenTree::Punct(ref p) => p.to_string() == punct,
-        _ => false,
+        input.parse::<Token![,]>()?;
+
+        Ok(DefaultReturn::Return(Some(return_closure)))
     }
 }
 
-enum TokenCheck {
-    UnexpectedToken(String, String),
-    UnexpectedEnd(String),
+pub(crate) struct Capture {
+    pub(crate) name: Expr,
+    pub(crate) alias: Option<Ident>,
+    pub(crate) kind: CaptureKind,
+    pub(crate) start: Span,
 }
 
-fn check_tokens(
-    tokens_to_check: &[SimpleToken],
-    parts: &mut PeekableProcIter,
-) -> Result<(), TokenCheck> {
-    let mut tokens = String::new();
-
-    for token in tokens_to_check {
-        if let Some(next) = parts.peek() {
-            if token != next {
-                return Err(TokenCheck::UnexpectedToken(tokens, next.to_string()));
-            }
-            tokens.push_str(token.to_str());
-            parts.next();
+impl Capture {
+    pub(crate) fn new(attr: &Attribute, input: ParseStream) -> syn::Result<Self> {
+        let kind = if attr.path().is_ident("strong") {
+            CaptureKind::Strong
+        } else if attr.path().is_ident("weak") {
+            CaptureKind::Weak
+        } else if attr.path().is_ident("weak_allow_none") {
+            CaptureKind::WeakAllowNone
+        } else if attr.path().is_ident("to_owned") {
+            CaptureKind::ToOwned
+        } else if attr.path().is_ident("watch") {
+            CaptureKind::Watch
         } else {
-            return Err(TokenCheck::UnexpectedEnd(tokens));
-        }
-    }
-    Ok(())
-}
+            unreachable!();
+        };
 
-#[doc(alias = "get_full_ident")]
-fn full_ident(
-    parts: &mut PeekableProcIter,
-    borrow_kind: BorrowKind,
-) -> Result<String, TokenStream> {
-    let mut name = String::new();
-    let mut prev_is_ident = false;
-
-    loop {
-        match parts.peek() {
-            Some(TokenTree::Punct(p)) => {
-                let p_s = p.to_string();
-                if p_s == "," || p_s == "=" {
-                    break;
-                } else if p_s == "." {
-                    if !prev_is_ident {
-                        return Err(parts.generate_error_with_next_span(&format!(
-                            "Unexpected `.` after `{}`",
-                            borrow_kind.to_str()
-                        )));
+        let mut alias = None;
+        if attr.meta.require_list().is_ok() {
+            attr.parse_nested_meta(|meta| {
+                if meta.path.is_ident("rename_to") {
+                    let value = meta.value()?;
+                    let lit = value.parse::<LitStr>()?;
+                    let id = lit.parse::<Ident>()?;
```

Comment:
Wouldn't it be simpler to ask for an ident directly?

```rust
#[rename_to = something]
```

doesn't seem that bad either.

Reply:
I didn't know that was possible. Better, thanks!